tests/test_logical_links.py

```diff
@@ -0,0 +1,147 @@
+        """
+        This test checks if the logic implemented for logically linked energy system, Meaning an
+        energy system where assets are directly connected to each other without a network
+        (transport asset(s)) in between.
+
+        Checks:
+        1. Check that energy is conserved
+
+        """
+        import models.unit_cases_gas.multi_demand_source_node.src.run_test as example
+        from models.unit_cases_gas.multi_demand_source_node.src.run_test import GasProblem
+
+        base_folder = Path(example.__file__).resolve().parent.parent
+
+        problem = run_esdl_mesido_optimization(
+            GasProblem,
+            esdl_file_name="test_logical_links.esdl",
+            base_folder=base_folder,
+            esdl_parser=ESDLFileParser,
+            profile_reader=ProfileReaderFromFile,
+            input_timeseries_file="timeseries.csv",
+        )
+        results = problem.extract_results()
+
+        np.testing.assert_allclose(
+            results["GasDemand_47d0.Gas_demand_mass_flow"]
+            + results["GasDemand_7978.Gas_demand_mass_flow"],
+            results["GasProducer_a977.Gas_source_mass_flow"]
+            + results["GasProducer_3573.Gas_source_mass_flow"],
+        )
+
+    def test_logical_links_nodes(self):
+
+        import models.unit_cases_gas.source_pipe_split_sink.src.run_source_sink as example
+        from models.unit_cases_gas.source_pipe_split_sink.src.run_source_sink import GasProblem
+
+        base_folder = Path(example.__file__).resolve().parent.parent
+
+        problem = run_esdl_mesido_optimization(
+            GasProblem,
+            esdl_file_name="split_with_logical_link.esdl",
+            base_folder=base_folder,
+            esdl_parser=ESDLFileParser,
+            profile_reader=ProfileReaderFromFile,
+            input_timeseries_file="timeseries.csv",
+        )
+        results = problem.extract_results()
+
+        # We check that no artificial gas in created
+        np.testing.assert_allclose(
+            results["GasProducer_0876.GasOut.Q"], results["GasDemand_a2d8.GasIn.Q"]
+        )
+
+        # We test conservaiton of flow at the nodes
+        for node, connected_pipes in problem.energy_system_topology.gas_nodes.items():
+            discharge_sum = 0.0
+
+            for i_conn, (_pipe, orientation) in connected_pipes.items():
+                discharge_sum += results[f"{node}.GasConn[{i_conn+1}].Q"] * orientation
+                np.testing.assert_allclose(
+                    results[f"{node}.GasConn[{i_conn+1}].H"], results[f"{node}.H"], atol=1.0e-6
+                )
+            np.testing.assert_allclose(discharge_sum, 0.0, atol=1.0e-12)
+
+    def test_logical_links_network_hybrid(self):
+        """
+        This test checks if the logic implemented for logically linked energy system, Meaning an
+        energy system where assets are directly connected to each other without a network
+        (transport asset(s)) in between.
```

Comment:
maybe state that hybrid refers to electricity and gas combined

Reply:
done